pkg/daemon/ceph/osd/volume.go

```diff
@@ -956,17 +956,28 @@ func GetCephVolumeRawOSDs(context *clusterd.Context, clusterInfo *client.Cluster
 				}
 			}
 			if encryptedBlock == "" {
-				// The encrypted block is not opened, this is an extreme corner case
-				// The OSD deployment has been removed manually AND the node rebooted
-				// So we need to re-open the block to re-hydrate the OSDInfo.
-				//
-				// Handling this case would mean, writing the encryption key on a temporary file, then call
-				// luksOpen to open the encrypted block and then call ceph-volume to list against the opened
-				// encrypted block.
-				// We don't implement this, yet and return an error.
-				return nil, errors.Errorf("failed to find the encrypted block device for %q, not opened?", block)
-			}
+				// The encrypted block is not opened.
+				// The OSD deployment has been removed manually.
```

Comment:
Why remove the mention of the node reboot? If the deployment is removed manually the encrypted block is left open.

Reply:
This is not clear, please clarify the scenario with CSI. If we have more than the reboot then they should all be listed in the code as comments.

Reply:
> This is not clear, please clarify the scenario with CSI. If we have more than the reboot then they should all be listed in the code as comments.

I dont think we can assume that a encrypted device opened by a pod remains open.

IMO The case of rook using disks directly on node/ intree driver provisioned pvc leaving the encrypted device open when 
the pod rebooted is the corner case. I am thinking we should go ahead with the assumption the encrypted device is no longer open once the pod, be it osd-prepare-job or osd,  is deleted.

Does this make sense?

Reply:
> > This is not clear, please clarify the scenario with CSI. If we have more than the reboot then they should all be listed in the code as comments.
> 
> I dont think we can assume that a encrypted device opened by a pod remains open after pod is deleted.

Why? What would close it then?

> 
> IMO The case of rook using disks directly on node/ intree driver provisioned pvc leaving the encrypted device open when the pod rebooted is the corner case. I am thinking we should go ahead with the assumption the encrypted device is no longer open once the pod, be it osd-prepare-job or osd, is deleted.

Even in the PVC case where Rook consumes a PVC for OSD disk, if that PVC is detached the encrypted block is still open and available but broken. We've had cases like this where the underlying block device disappears.

> 
> Does this make sense?

Reply:
> > > This is not clear, please clarify the scenario with CSI. If we have more than the reboot then they should all be listed in the code as comments.
> > 
> > 
> > I dont think we can assume that a encrypted device opened by a pod remains open after pod is deleted.
> 
> Why? What would close it then?

Unmounting of PVC from the node closes it always?

> 
> > IMO The case of rook using disks directly on node/ intree driver provisioned pvc leaving the encrypted device open when the pod rebooted is the corner case. I am thinking we should go ahead with the assumption the encrypted device is no longer open once the pod, be it osd-prepare-job or osd, is deleted.
> 
> Even in the PVC case where Rook consumes a PVC for OSD disk, if that PVC is detached the encrypted block is still open and available but broken. We've had cases like this where the underlying block device disappears.

csi managed PVC are always unmounted when no pod is using that PVC, this will be the normal case going forward ?

Reply:
> > > > > This is not clear, please clarify the scenario with CSI. If we have more than the reboot then they should all be listed in the code as comments.
> > > > 
> > > > 
> > > > I dont think we can assume that a encrypted device opened by a pod remains open after pod is deleted.
> > > 
> > > 
> > > Why? What would close it then?
> > 
> > 
> > Unmounting of PVC from the node closes it always?
> 
> No, detaching the underlying device and even force detach won't close the encrypted block, the dm would still be present.

I've added these cases in the comments, so it should be fine for now. 

As far as my testing goes, the dm device is gone, 

simplest way I am testing is using ceph cluster on rbd PVC :sweat_smile: 
just enough to simulate csi based pvc environment
I've seen this on aws gp2-csi and vsphere csi provisioner too.

Reply:
Sounds good 👍🏻 I don't mean to argue too much too, if you saw it then I trust you. Also as a reference see this code https://github.com/leseb/rook/blob/61279e98201d5d80090f57433a21e297b2f5efb5/pkg/operator/ceph/cluster/osd/spec.go#L172-L212. We still have cases where the PV block is detached and the encrypted block is still around.

Reply:
okay :+1: 
one example is as follows:

This is how it looks on a node with functioning osd
```
[rakshith@fedora ~]$ oc debug node/ip-10-0-165-109.us-east-2.compute.internal
Starting pod/ip-10-0-165-109us-east-2computeinternal-debug ...
To use host binaries, run `chroot /host`
Pod IP: 10.0.165.109
If you don't see a command prompt, try pressing enter.
sh-4.4# chroot /host
sh-4.4# lsblk
NAME        MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
loop1         7:1    0   100G  0 loop  
rbd0        252:0    0    40G  0 disk  /var/lib/kubelet/pods/d3adec98-e695-4ba3-a110-ff654c5a5f5c/volumes/kubernetes.io~csi/pvc-a3508c23-c949-43bd-bbfe-e00b52
nvme0n1     259:0    0   120G  0 disk  
|-nvme0n1p1 259:1    0     1M  0 part  
|-nvme0n1p2 259:2    0   127M  0 part  
|-nvme0n1p3 259:3    0   384M  0 part  /boot
`-nvme0n1p4 259:4    0 119.5G  0 part  /sysroot
nvme1n1     259:5    0    50G  0 disk  /var/lib/kubelet/pods/2e1983d0-25cf-436c-b289-ce7acc6625c4/volumes/kubernetes.io~csi/pvc-0a2fa1bb-6af1-4ca1-a624-12d9f0
nvme2n1     259:6    0   100G  0 disk  
`-ocs-deviceset-0-data-0cljr5-block-dmcrypt
            253:0    0   100G  0 crypt 
```

This is how it looks when osd removal is done 
```
[rakshith@fedora ~]$ oc debug node/ip-10-0-152-142.us-east-2.compute.internal
Starting pod/ip-10-0-152-142us-east-2computeinternal-debug ...
To use host binaries, run `chroot /host`
Pod IP: 10.0.152.142
If you don't see a command prompt, try pressing enter.
sh-4.4# 
sh-4.4# lsblk
NAME        MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
rbd0        252:0    0    40G  0 disk /host/var/lib/kubelet/pods/31cdabdd-9f4a-4490-b8a2-cd9a9e1f1458/volumes/kubernetes.io~csi/pvc-76c14ea6-8664-4116-b86e-80a
nvme0n1     259:0    0   120G  0 disk 
|-nvme0n1p1 259:1    0     1M  0 part 
|-nvme0n1p2 259:2    0   127M  0 part 
|-nvme0n1p3 259:3    0   384M  0 part /host/boot
`-nvme0n1p4 259:4    0 119.5G  0 part /host/sysroot
nvme1n1     259:5    0    50G  0 disk /host/var/lib/kubelet/pods/46933254-64a2-43a9-a5b5-d4904ccc2025/volumes/kubernetes.io~csi/pvc-95b398fd-3d76-456d-903f-8bc
```

When osd-prepare job pod starts
```
sh-4.4# lsblk
NAME        MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
loop0         7:0    0   100G  0 loop 
rbd0        252:0    0    40G  0 disk /var/lib/kubelet/pods/31cdabdd-9f4a-4490-b8a2-cd9a9e1f1458/volumes/kubernetes.io~csi/pvc-76c14ea6-8664-4116-b86e-80abcf33
nvme0n1     259:0    0   120G  0 disk 
|-nvme0n1p1 259:1    0     1M  0 part 
|-nvme0n1p2 259:2    0   127M  0 part 
|-nvme0n1p3 259:3    0   384M  0 part /boot
`-nvme0n1p4 259:4    0 119.5G  0 part /sysroot
nvme1n1     259:5    0    50G  0 disk /var/lib/kubelet/pods/46933254-64a2-43a9-a5b5-d4904ccc2025/volumes/kubernetes.io~csi/pvc-95b398fd-3d76-456d-903f-8bca598f
nvme3n2     259:7    0   100G  0 disk 
sh-4.4# exit
```

Reply:
Ok but `ocs-deviceset-0-data-0cljr5-block-dmcrypt` is copied in `/var/lib/ceph/osd/ceph-0/block`, this is the file the OSD uses to run and I believe this file is still there.

Reply:
> Ok but `ocs-deviceset-0-data-0cljr5-block-dmcrypt` is copied in `/var/lib/ceph/osd/ceph-0/block`, this is the file the OSD uses to run and I believe this file is still there.

that is a file inside the container right? 
it goes away with the deletion of osd pod?

Reply:
No this is HostPath.